node/cli/src/service.rs

```diff
@@ -110,44 +113,40 @@ construct_service_factory! {
 					service.spawn_task(Box::new(select));
 				}
 
-				let grandpa_key = if service.config.disable_grandpa {
-					None
-				} else {
-					service.authority_key::<grandpa_primitives::AuthorityPair>()
-				};
-
 				let config = grandpa::Config {
-					local_key: grandpa_key.map(Arc::new),
 					// FIXME #1578 make this available through chainspec
 					gossip_duration: Duration::from_millis(333),
 					justification_period: 4096,
 					name: Some(service.config.name.clone())
 				};
 
-				match config.local_key {
-					None if !service.config.grandpa_voter => {
-						service.spawn_task(Box::new(grandpa::run_grandpa_observer(
-							config,
-							link_half,
-							service.network(),
-							service.on_exit(),
-						)?));
-					},
-					// Either config.local_key is set, or user forced voter service via `--grandpa-voter` flag.
-					_ => {
-						let telemetry_on_connect = TelemetryOnConnect {
-							telemetry_connection_sinks: service.telemetry_on_connect_stream(),
-						};
-						let grandpa_config = grandpa::GrandpaParams {
-							config: config,
-							link: link_half,
-							network: service.network(),
-							inherent_data_providers: service.config.custom.inherent_data_providers.clone(),
-							on_exit: service.on_exit(),
-							telemetry_on_connect: Some(telemetry_on_connect),
-						};
-						service.spawn_task(Box::new(grandpa::run_grandpa_voter(grandpa_config)?));
-					},
+				if service.config.roles == Roles::AUTHORITY && !service.config.disable_grandpa {
```

Comment:
This is not correct, you might want to run the grandpa voter as a full node (in fact in master we just added a flag for this).

Reply:
Is there a reason why you can't just use --validator? @mxinden thought that would also work for sentry, but I can add it back if you need a separate flag.

Reply:
If aura/babe are not staked as validators they won't author a block, since they don't get a slot, it's harmless to run the aura task, are you concerned about performance?

Reply:
ok I think the best and clearest solution is having an `--aura` `--grandpa-voter` and `--validator` option where `--validator` is just a shortcut for --aura --grandpa-voter. Then we can support all use cases and the semantics are clear (opt-in)